File: Python_Scripting_for_Computational_Science_Third_Edition/py/regex/swap3.py
#!/usr/bin/env python
import re, glob, string
# as swap1.py except that we here use a regex with comments and compile it:
arg = r'[^,]+'
call = re.compile(r"""
         (?P<start>    # preserve start of function string
         superLibFunc  #   name of function to match 
         \s*           #   optional whitespace
         \(            #   parenthesis before argument list
         \s*           #   optional whitespace
         )             # end of <start>
         (?P<arg1>.+?) # a C variable name, non-greedy
         (?P<middle>   # preserve middle of function string
         \s*,\s*       #   comma with optional surrounding whitespace
         )             # end of <middle>
         (?P<arg2>.+?) # a C variable name, non-greedy
         (?P<end>      # preserve end of function string
         \s*           #   optional whitespace
         \)            #   closing parenthesis
         )             # End of <end>
         """ , re.VERBOSE | re.DOTALL)

cfiles = ['.test1.c']
for cfile in cfiles:
    print 'Treating',cfile
    file = open(cfile, 'r')
    filestr = file.read() # load all lines into a string
    file.close()
    filestr = call.sub(r'superLibFunc(\g<arg2>\g<middle>\g<arg1>\g<end>)',
                       filestr)
    file = open(cfile + '.tmp', 'w')
    file.write(filestr)  # print everything into cfile.tmp

File: Python_Scripting_for_Computational_Science_Third_Edition/perl/swap3.pl
#!/usr/bin/perl
# as swap1.pl but with comments inside regular expressions
# as swap3.pl but not with the regex in a string

$arg  = "[^,]+";

@cfiles = ("../py/regex/.test1.c");
foreach $cfile (@cfiles) {
    print "Treating $cfile...\n";
    open(FILE,"<$cfile") || die "Problems: $!\n";
    @lines = <FILE>;  # load all lines into a list
    close(FILE);
    $filestr = join("",@lines);
    $filestr =~ s{
        superLibFunc     # name of function to match 
           \s*           # possible whitespace
           \(            # left parenthesis
           \s*           # possible whitespace
           ($arg)        # first argument plus optional whitespace
           ,             # comma between the arguments
           \s*           # possible whitespace
           ($arg)        # second argument plus optional whitespace
           \)            # closing parenthesis
    }{superLibFunc($2, $1)}gx;
    open(FILE,">$cfile.tmp");
    print FILE $filestr;
}
